shnet_main.c

```diff
@@ -396,12 +382,50 @@ ssize_t shnet_port_read(struct file *file, char __user *buf, size_t size,
 	return sz;
 }
 
+ssize_t shnet_port_write(struct file *file, const char __user *buf, size_t size,
+			 loff_t *ppos)
+{
+	int rc, sz = 0;
+	struct shnet_req req;
+
+	while (1) {
+		if (sz + sizeof(struct shnet_req) > size)
+			goto out;
+
+		rc = copy_from_user(&req,
+				    (struct shnet_req __user *)(buf + sz),
+				    sizeof(req));
+		if (rc) {
+			pr_err("Fail to copy from user buf, pos=%d\n", sz);
+			return sz;
```

Comment:
But what about all the messages that already processed? i don't think it is a good practice to fail the entire transaction for only one buggy message. Consider a process that prepares an array of 100 entries, as far as i know the write call should return the number of bytes that were processed, the caller then need to figure out what went wrong.